### DISCUSSION TITLE
The creation date of the date plug-in seems to be incorrectly displayed

### DISCUSSION THREAD

**Discussion by w1ndys** [squidfunk/mkdocs-material#7161] [Q&A] [answered]
After raising the https://github.com/squidfunk/mkdocs-material/discussions/7135 issue again, I took the official approach and found a problem, which was not solved by looking through the community.

I uploaded a new file (2024-05-06). The creation date should be 2024-05-06, but the actual page display is earlier.

Here are some reference screenshots and links

![Screenshot_2024-05-07-10-43-21-500_com android chrome-edit](https://github.com/squidfunk/mkdocs-material/assets/109416673/8883b9fa-178b-4d14-97e7-dd6907f48755)

![Screenshot_2024-05-07-10-39-17-450_com android chrome](https://github.com/squidfunk/mkdocs-material/assets/109416673/c66c53c4-e82b-49e1-b5d3-27ebedcb573b)

![Screenshot_2024-05-07-10-39-14-262_com android chrome](https://github.com/squidfunk/mkdocs-material/assets/109416673/0cba48a1-d983-4ba7-8b3e-83398cc05c79)

Repository file：https://github.com/W1ndys/Easy-QFNU/blame/main/docs/Easy-College/index.md

The API link of this file: https://api.github.com/repos/W1ndys/Easy-QFNU/commits?path=docs/Easy-College/index.md

Actual front-end page address: https://easy-qfnu.top/Easy-College/

**Comment by w1ndys** [comment]
I found that not only this file, but also other files have this kind of situation, and the fetch-depth of the plug-in has been set to 0.🤔🤔🤔

The date of modification should be correct, but the date of creation is wrong.

**Reply by kamilkrzyskow** [reply]
Hello @W1ndys,
unfortunately, your git history got a bit "corrupted" over the time. Here is the code used by the `mkdocs-git-revision-date-localized-plugin` to get the creation timestamp:
https://github.com/timvink/mkdocs-git-revision-date-localized-plugin/blob/b0ab74f9c4cba65f25cc3a28cd5f3234f297b5c0/mkdocs_git_revision_date_localized_plugin/util.py#L83-L92
```python
commit_timestamp = git.log(
    realpath, date="unix", format="%at", diff_filter="A", no_show_signature=True, follow=True
)
```
which partially translates to this command:
```shell
git log --diff-filter="A" --follow docs/Easy-College/index.md
```

<details>
<summary> Result </summary>

```shell
commit 84b735d33a3b569e22f7e7ea1b2cca90b1a4417c
Author: W1ndys <email>
Date:   Fri Apr 19 11:31:51 2024 +0800

    恢复调整的目录文件

commit 71ba1a9dec9770372528bc4fc02bdffd5a2f7402
Author: W1ndys <email>
Date:   Thu Mar 7 15:05:45 2024 +0800

    若干配置
```

</details>

and the plugin uses the last found commit of the `--follow` option. 

---

However, this command doesn't help much with debugging the issue, so I did a quick search online for alternatives and found this:
https://kgrz.io/use-git-log-follow-for-file-history.html
```shell
git log --oneline -M --stat --follow -- docs/Easy-College/index.md
```

<details>
<summary> Result </summary>

```shell
931f4b1 调整Easy-学院的布局，新增Easy-College上层目录
 docs/{Cyber/Organization/Lab => Easy-College}/index.md | 10 +++++-----
 1 file changed, 5 insertions(+), 5 deletions(-)
e627272 更新格式
 docs/Cyber/Organization/Lab/index.md | 1 -
 1 file changed, 1 deletion(-)
9970f02 Update index.md
 docs/Cyber/Organization/Lab/index.md | 6 ------
 1 file changed, 6 deletions(-)
6fb7fcb 每日定时自动更新文档日期
 docs/Cyber/Organization/Lab/index.md | 19 ++++++++++++-------
 1 file changed, 12 insertions(+), 7 deletions(-)
ce0540b 删除所有日期
 docs/Cyber/Organization/Lab/index.md | 5 +----
 1 file changed, 1 insertion(+), 4 deletions(-)
e83d20f 刷新日期2024年4月28日
 docs/Cyber/Organization/Lab/index.md | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)
22a6c1e 网安信息填充
 docs/Cyber/Organization/Lab/index.md | 10 ++++++++++
 1 file changed, 10 insertions(+)
5c28ba8 规范网安的路径
 docs/{javascripts/extra.js => Cyber/Organization/Lab/index.md} | 0
 1 file changed, 0 insertions(+), 0 deletions(-)
84b735d 恢复调整的目录文件
 docs/javascripts/extra.js | 0
 1 file changed, 0 insertions(+), 0 deletions(-)
71e40c3 删除需要调整的目录文件
 docs/javascripts/extra.js | 0
 1 file changed, 0 insertions(+), 0 deletions(-)
71ba1a9 若干配置
 docs/javascripts/extra.js | 0
 1 file changed, 0 insertions(+), 0 deletions(-)
```

</details>

In summary in commit `71ba1a9` you created an empty `docs/javascripts/extra.js` file, and later on in commit `5c28ba8` git detected a rename of the empty `docs/javascripts/extra.js` to `docs/Cyber/Organization/Lab/index.md`, and later on it detected a rename from `docs/Cyber/Organization/Lab/index.md` to `docs/Easy-College/index.md`. So the creation date is used from the first `extra.js` file.

So this is an unfortunate case of git history breakage, you can't really do anything about it 🤔, modifying the commits and then doing a force push would only lead to more issues.

Perhaps you could ask upstream, for a new option in the plugin which would allow to control the usage of `--follow` flag.

**Reply by w1ndys** [reply]
> and later on in commit 5c28ba8 git detected a rename of the empty docs/javascripts/extra.js to docs/Cyber/Organization/Lab/index.md, and later on it detected a rename from docs/Cyber/Organization/Lab/index.md to docs/Easy-College/index.md.

I don't know why the git history is inconsistent with my actual operation.🤔 I think your judgment is correct, but the git history error shows that I have all operated, I don't understand why git history shows up like this.
My actual operation: Created `extra.js`, files such as `docs/Cyber/.. `were created at some point, and files such as `docs/Cyber/..`  were moved `docs/Easy-College/Cyber/..`

---

According to what you said, if I can't do anything about the plug-in settings and can only solve the problem by myself, should I try to avoid the behavior of' moving Files, instead, create a new identical file and delete the original file. Is this correct?

In other words, how should I avoid this destructive behavior?😳

Do you have a better solution?🤔🤔🤔
I am grateful for any help.

**Reply by kamilkrzyskow** [reply, marked as answer]
> I don't understand why git history shows up like this.

I don't know either, but you add a lot of empty files to the git repository. Based on the last command they even changed, despite being still empty, so perhaps your code editor changes the encoding of the files and that gets somehow stored in the git history 🤔 

> should I try to avoid the behavior of' moving Files, instead, create a new identical file and delete the original file. Is this correct?

Adding and deleting a file won't help, I tried it, so unless I did something wrong I validated my claim. 
The `--diff-filter="A" --follow <PATH>` still finds the old creation commits, which are falsely detected based on the falsely detected renames. So basically you can't fix the git history by adding stuff, you'd need to edit previous commits and force push to the repository. 
But like I said before, better not to do it as other dates could become corrupt through this way, and forks will be incompatible.

> According to what you said, if I can't do anything about the plug-in settings

Currently not, but I said you could ask the author of the plugin to enable such configuration, although I don't know what's the best course of action here, as your case seems to be quite unique 🤔 

I think the best approach would be to just write a Python hook that would edit the `page.meta.git_creation_date_localized` value for the affected files with the corrupted dates. You could combine your previous approach from https://github.com/squidfunk/mkdocs-material/discussions/7135 together with the code from the plugin to get the "most recent" creation date.

**Reply by w1ndys** [reply]
It sounds very good. I have already raised [issues](https://github.com/timvink/mkdocs-git-revision-date-localized-plugin/issues/134#/) with the plug-in. Now we just need to wait for the official reply from the plug-in author.

Before the official fixes this "error", I think I can temporarily turn off the display of "Creation Date", because this is not a fatal error, it just affects the viewing experience and that's all🫡🫡🫡

**Reply by timvink** [reply]
Wow great analysis done into this edge case!

It seems to be exactly what is described here: https://blog.plover.com/prog/git-log-follow.html

That's from 2019. Maybe there has been updates to git or new flags that haven't been tried. I went through the git logs docs (https://www.git-scm.com/docs/git-log/2.11.4) and there are some options to try:

--ignore-all-space
--ignore-blank-lines

And maybe the options --ignore-missing and -L (limit line range to first line and after to avoid the empty files)

Can you try this out on your repo? Happy to update the plugin with an improved version of the git command for finding creation dates.

**Reply by w1ndys** [reply]
After testing, it was found that only adding the `--follow` parameter will get wrong results.
By default, adding the `ignore` parameter will not give incorrect results.
The following is the running log
Thank you for your help

```
W1ndys@W1ndys MINGW64 /Easy-qfnu (main)
$ git log docs/Easy-College/index.md
commit 931f4b1d36e2c8d1e3f4a29b94e18a174bd5d29e
Author: W1ndys <w1ndys@outlook.com>
Date:   Mon May 6 19:20:36 2024 +0800

    调整Easy-学院的布局，新增Easy-College上层目录

W1ndys@W1ndys MINGW64 /Easy-qfnu (main)
$ git log --ignore-all-space  docs/Easy-College/index.md
commit 931f4b1d36e2c8d1e3f4a29b94e18a174bd5d29e
Author: W1ndys <w1ndys@outlook.com>
Date:   Mon May 6 19:20:36 2024 +0800

    调整Easy-学院的布局，新增Easy-College上层目录

W1ndys@W1ndys MINGW64 /Easy-qfnu (main)
$ git log --ignore-blank-lines   docs/Easy-College/index.md
commit 931f4b1d36e2c8d1e3f4a29b94e18a174bd5d29e
Author: W1ndys <w1ndys@outlook.com>
Date:   Mon May 6 19:20:36 2024 +0800

    调整Easy-学院的布局，新增Easy-College上层目录

W1ndys@W1ndys MINGW64 /Easy-qfnu (main)
$ git log --follow    docs/Easy-College/index.md
commit 931f4b1d36e2c8d1e3f4a29b94e18a174bd5d29e
Author: W1ndys <w1ndys@outlook.com>
Date:   Mon May 6 19:20:36 2024 +0800

    调整Easy-学院的布局，新增Easy-College上层目录

commit e6272721b58450f9b47530a36592a6bdfadb655e
Author: W1ndys <w1ndys@outlook.com>
Date:   Sun May 5 17:19:27 2024 +0800

    更新格式

commit 9970f02ba13a818ebc7cdb195006197411bc367a
Author: W1ndys <w1ndys@Outlook.com>
Date:   Wed May 1 09:22:06 2024 +0800

    Update index.md

.
.
.
.
.
.
.
.
More irrelevant information has been omitted
```

**Reply by w1ndys** [reply]
Is it enough to turn off the follow below?
```
commit_timestamp = git.log(
     realpath, date="unix", format="%at", diff_filter="A", no_show_signature=True, follow=True
)
```
to
```
commit_timestamp = git.log(
     realpath, date="unix", format="%at", diff_filter="A", no_show_signature=True, follow=False
)
```

**Reply by timvink** [reply]
Ok that's progress. But we need --follow, because some users had already requested it earlier (file renames are quite common).

So git log --follow --ignore-blank-lines does not work for you either?

**Reply by w1ndys** [reply]
I'm afraid the phenomenon is not satisfactory. I just tested and found that
```
git log --follow --ignore-blank-lines
git log --follow --ignore-all-space
```
Both commands will get wrong information. I guess as long as there is `follow`, it will go wrong. It seems that this is a conflict problem.

**Reply by timvink** [reply]
There is no single solution then. In your case, you don't want to use follow renames, but others do. Most users don't commit empty files, so your case is going to be rarer.

It's possible to add an option to the plugin to specify whether to follow or not. Would that fix it for you?

**Reply by w1ndys** [reply]
great!
I think this should solve my problem
It is worth mentioning that I actually don’t know the specific reason for the error in the Git history. As of now, I am not sure that the reason is because I submitted an empty file.
But I think your idea can solve my problem. The key to solving my problem is whether to execute `follow`
thank you very much for your help

**Reply by timvink** [reply]
FYI, the plugin now uses `git --follow --ignore-blank-lines --ignore-all-space` by default, and has a new option to disable `--follow`. 

See the release notes: https://github.com/timvink/mkdocs-git-revision-date-localized-plugin/releases/tag/v1.3.0
